c++/src/OrcHdfsFile.cc

```diff
@@ -0,0 +1,170 @@
+
+  class HdfsFileInputStream : public InputStream {
+  private:
+    std::string filename;
+    std::unique_ptr<hdfs::FileHandle> file;
+    std::shared_ptr<hdfs::FileSystem> file_system;
+    uint64_t totalLength;
+
+  public:
+    HdfsFileInputStream(std::string _filename) {
+      filename = _filename ;
+
+      //Building a URI object from the given uri_path
+      hdfs::optional<hdfs::URI> uri = hdfs::URI::parse_from_string(filename);
+      if (!uri) {
+        throw ParseError("Malformed URI: " + filename);
+      }
+
+      hdfs::Options options;
+      //Setting the config path to the default: "$HADOOP_CONF_DIR" or "/etc/hadoop/conf"
+      hdfs::ConfigurationLoader loader;
+      //Loading default config files core-site.xml and hdfs-site.xml from the config path
+      hdfs::optional<hdfs::HdfsConfiguration> config = loader.LoadDefaultResources<hdfs::HdfsConfiguration>();
+      //TODO: HDFS-9539 - after this is resolved, valid config will always be returned.
+      if(config){
+        //Loading options from the config
+        options = config->GetOptions();
+      }
+      hdfs::IoService * io_service = hdfs::IoService::New();
+      //Wrapping fs into a shared pointer to guarantee deletion
+      std::shared_ptr<hdfs::FileSystem> fs(hdfs::FileSystem::New(io_service, "", options));
+      if (!fs) {
+        throw ParseError("Can't create FileSystem object. ");
+      }
+      hdfs::Status status;
+      //Check if the user supplied the host
+      if(!uri.value().get_host().empty()){
+        //If port is supplied we use it, otherwise we use the empty string so that it will be looked up in configs.
+        std::string port = (uri.value().get_port()) ? std::to_string(uri.value().get_port().value()) : "";
+        status = fs->Connect(uri.value().get_host(), port);
+        if (!status.ok()) {
+          throw ParseError("Can't connect to " + uri.value().get_host() + ":" + port + ". " + status.ToString());
+        }
+      } else {
+        status = fs->ConnectToDefaultFs();
+        if (!status.ok()) {
+          if(!options.defaultFS.get_host().empty()){
+            throw ParseError("Error connecting to " + options.defaultFS.str() + ". " + status.ToString());
+          } else {
+            throw ParseError("Error connecting to the cluster: defaultFS is empty. " + status.ToString());
+          }
+        }
+      }
+
+      if (!fs) {
+        throw ParseError("Can't connect the file system. ");
+      }
+
+      hdfs::FileHandle *file_raw = nullptr;
+      status = fs->Open(uri->get_path(), &file_raw);
+      if (!status.ok()) {
+        throw ParseError("Can't open " + uri->get_path() + ". " + status.ToString());
+      }
+      //wrapping file_raw into a unique pointer to guarantee deletion
+      std::unique_ptr<hdfs::FileHandle> file_handle(file_raw);
+      file = std::move(file_handle);
+      //transferring the ownership of FileSystem to HdfsFileInputStream to avoid premature deallocation
+      file_system = fs;
```

Comment:
Not quite understand why we need a local variable "fs". You could just initialize this->file_system. Also "fs" should have been a unique_ptr as it doesn't share ownership with anyone.